### ISSUE TITLE
Extend CoreFn's Int literal encoding with a stringified value

### ISSUE THREAD

**Issue by natefaubion** [purescript/purescript#4506]
## Summary

CoreFn's integer encoding encodes the `Integer` (unbounded) as a JSON Number literal. This is technically valid as JSON does not require Number bounds, but makes tooling across languages complicated. For example, stock JSON tooling for JS will only support a 64bit double (52 bit max integer), and anything outside that range is difficult to parse.

## Motivation

It's not possible for tools like `purescript-backend-optimizer` to parse CoreFn JSON for backends which use unbounded integers with any of the stock JSON tooling available in the PureScript ecosystem as they all rely on builtin `JSON.parse`.

## Proposal

https://github.com/purescript/purescript/blob/8ede652b6f040dcaadff8a59f9cfb118993f7986/src/Language/PureScript/CoreFn/ToJSON.hs#L63-L66

Extend this with a `stringValue` field which prints the value as a String, making it easier for tooling to parse this as a BigInt of some kind. The additional field is necessary to keep CoreFn backwards compatible.

**Comment by sigma-andex** [comment]
> It's not possible for tools like `purescript-backend-optimizer` to parse CoreFn JSON for backends which use unbounded integers with any of the stock JSON tooling available in the PureScript ecosystem as they all rely on builtin `JSON.parse`.

Not sure if this helps you, but [`yoga-json`](https://github.com/rowtype-yoga/purescript-yoga-json) is able to parse `BigInt` by passing a reviver function to `JSON.parse`.

**Comment by natefaubion** [comment]
> Not sure if this helps you, but [yoga-json](https://github.com/rowtype-yoga/purescript-yoga-json) is able to parse BigInt by passing a reviver function to JSON.parse.

Reviver functions don’t help here because it’s already a JSON Number. Reviver functions can’t recover the original JSON syntax in order to load a Number as a BigInt without losing precision.

**Comment by michaelficarra** [comment]
@natefaubion You may be interested in the Stage 3 proposal [JSON.parse source text access](https://github.com/tc39/proposal-json-parse-with-source) which makes this possible through revivers. Stage 3 means recommended for implementation, and it is [already available in Chrome stable](https://test262.fyi/#staging/JSON|v8_exp,sm_exp,jsc,qjs,libjs,chakra,graaljs,xs,rhino,babel).

**Comment by garyb** [comment]
Very cool!

**Comment by JordanMartinez** [comment]
> @natefaubion You may be interested in the Stage 3 proposal [JSON.parse source text access](https://github.com/tc39/proposal-json-parse-with-source) which makes this possible through revivers. Stage 3 means recommended for implementation, and it is [already available in Chrome stable](https://test262.fyi/#staging/JSON%7Cv8_exp,sm_exp,jsc,qjs,libjs,chakra,graaljs,xs,rhino,babel).

I still think it would be easier to just encode the value as a String.

**Comment by natefaubion** [comment]
I don't think so. Number values are `Double` which are IEEE floats, same as JS, so there's no precision issue.

**Comment by anttih** [comment]
This might be a slightly different incarnation of this issue I just came across while compiling the `integers` test suite using `backend-optimizer` (thus using corefn as input). The PS literal `-2147483648` gets compiled to `(negate 2147483648)`. That positive number literal doesn't fit the 32-bit Integer once it's converted to `int` with `2147483648|0`. Changing the JSON representation to a string wouldn't actually help if the negative literal gets compiled to `negate x`, the damage is already done in the corefn.
